test/jackdaw/word_count_test.clj

```diff
@@ -0,0 +1,160 @@
+(ns jackdaw.word-count-test
+  (:require [clojure.test :refer :all]
+            [jackdaw.streams :as k]
+            [jackdaw.test :as jd.test]
+            [jackdaw.test.fixtures :as fix])
+  (:import org.apache.kafka.common.serialization.Serdes))
+
+;;; Example of using the test-machine ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; Here we take the word-count example from confluent's examples repo
+;;
+;;   https://github.com/confluentinc/kafka-streams-examples
+;;
+;; and provide an implementation in Clojure, and test it using the
+;; test-machine.
+;;
+;; First the app. All pretty simple stuff.
+;;
+;;   - Parse each line into a list of words
+;;   - Group By each unique word
+;;   - Count the records in each group
+;;   - Write the counts to the output topic
+
+(defn parse-line
+  [line]
+  (let [line (-> (.toLowerCase line (java.util.Locale/getDefault)))]
+    (->> (.split line "\\W+")
+         (into []))))
+
+(defn word-count
+  [in out]
+  (fn [builder]
+    (let [counts (-> (k/kstream builder in)
+                     (k/flat-map-values (fn [line]
+                                          (parse-line line)))
+                     (k/group-by (fn [[k v]]
+                                   v))
+                     (k/count)
+                     (k/to-kstream))]
+      (k/to counts out)
+      builder)))
+
+;; In order to use the app builder defined above, we need to provide
+;; it with topic definitions for the input/output topics. In a real app, you
+;; might consider loading this type of configuration from an EDN file
+
+(def input
+  {:topic-name "streams-plaintext-input"
+   :replication-factor 1
+   :partition-count 1
+   :key-serde (Serdes/String)
+   :value-serde (Serdes/String)})
+
+(def output
+  {:topic-name "streams-plaintext-output"
+   :replication-factor 1
+   :partition-count 1
+   :key-serde (Serdes/String)
+   :value-serde (Serdes/Long)})
+
+(def word-count-topics
+  {:input input
+   :output output})
+
+;; For testing, we need to configure a few elements
+;;
+;;  - the "broker config" is shared by both the app and test configs and is
+;;    needed to connect to kafka
+;;  - the "app config" is a StreamsConfig and is supplied when building
+;;    the app topology
+;;  - the "test config" is a ConsumerConfig and is supplied when creating the
+;;    consumer used by the test machine
+
+(def broker-config
+  {"bootstrap.servers" "localhost:9092"})
+
+(def app-config
+  (assoc broker-config
+   "cache.max.bytes.buffering" "0"
+   "application.id" "streams-word-count"
+   "default.key.serde" "org.apache.kafka.common.serialization.Serdes$StringSerde"
+   "default.value.serde" "org.apache.kafka.common.serialization.Serdes$StringSerde"))
+
+(def test-config
+  (assoc broker-config
+    "group.id" "word-count-test"))
+
+;; Now we're ready to build a test
+
+(defn write-lines
```

Comment:
This function doesn't actually write lines, does it? Should we maybe name it after the data returned, for example, `write-commands`? 

The same could be said of `jackdaw.word-count-test/watch-for-output` below.

Reply:
These are just private test helpers so I don't think the names are hugely important but I think that `write-commands` is worse. I tried to choose names for how they will eventually be used in the context of a test. The "lines" bit is intended to make the language of the test match the language of problem being solved.

```
      (let [{:keys [results journal]} (->> (concat
                                            (write-lines)
                                            (watch-for-output "understand"))
                                           (jd.test/run-test machine))
```

Within the context of the `concat`, tests will often use helpers to build up commands. It would get pretty tedious if every function had the word "commands" in it.